docs/internals.md

```diff
@@ -1,10 +1,433 @@
+  - (W) afterBuild      : Executed only once after the `WidgetState` was created. Is executed *after* all default values have been applied.
+  - (W) connectEvents   : Only relevant for renderables. Executed every time a callback is supposed to be attached to the underlying GTK widget. It defines how to do so.
+  - (W) disconnectEvents: Only relevant for renderables. Executed every time a callback is supposed to be removed from the underlying GTK widget. It defines how to do so.
+  - (WF) update         : Executed every time `WidgetState` is updated by `Widget`.
+  - (F) property        : Executed every time the hook-field changed its value in both the update and build phases.
+  - (F) read            : Used in `Dialog`. Executes every time the state of the underlying GTK-Widget changes.
+
+W: Can act as hook for Widgets
+F: Can act as hook for fields
+
+All hooks have implicit access to a variable called `state`, which contains the `WidgetState`-instance of your particular widget.
+
+With the exception of `read`, all hooks also have implicit access to a variable called `widget`, which is the `Widget` instance.
+
+Generally the `build` and `update` hook are likely to have the highest utility for you. Consult their individual sections for more information.
+
+#### Build Hook
+The `build` hook runs just before any values are assigned to the `WidgetState`.
+
+The intended usecase for build-hooks is adding logic that sets fields on `WidgetState` that don't have default-values.
+
+Example: A Widget may need to load data from elsewhere, via a file or HTTP request. Doing this in the `view` method would rerun the code every time the widget re-renders. But in the build-hook, it only runs once during widget-creation. 
+
+Here a simple code-example:
+
+```nim
+import std/json
+
+type Config = object
+  name: string
+
+viewable MyViewable:
+  config: Config
+    
+  hooks:
+    build:
+      state.config = "./src/bla.json".readFile().parseJson().to(Config)
+
+method view(state: MyViewableState): Widget =
+  result = gui:
+    Label:
+      text = state.config.name
+
+
+## The App
+viewable App:
+  discard
+
+method view(app: AppState): Widget =
+  result = gui:
+    Window:
+      MyViewable()
+
+brew(gui(App()))
+```
+
+#### Before-Build Hook
+The `beforeBuild` hook runs before the build-hook and thus also before any values are assigned to the `WidgetState`.
+
+Their main usecase is renderables, where they are used to instantiate the GTK-Widget and assign it to `internalWidget` on `WidgetState`.
+
+Here a simple code-example:
+
+```nim
+import owlkettle
+import owlkettle/gtk
+import std/json
+
+type Config = object
+  name: string
+
+renderable MyRenderable:
+  text: string
+    
+  hooks:
+    beforeBuild:
+      echo state.repr
+      state.internalWidget = gtk_label_new("")
+
+    afterBuild:
+      gtk_label_set_text(state.internalWidget, state.text.cstring)
```

Comment:
This should probably use a property hook. You could also set a static text in `gtk_label_new` to simplify the example even more.

Reply:
I update this section to ineed just do `gtk_label_new`, because that reduces the noise as much as possible.

As for `afterBuild`, I reused that example in the `afterBuild` section, since it provides a better example for what `afterBuild` can be useful for than what stood there previously (see line 390 and following). It's still with the same flaw (So if you update the parent this child will not change), but imo that makes sense if you want a renderable and only set its values once. It provides a nice guarantee that these "immutable" widgets in a sense are not going to do strange stuff on you regardless of how the AppState changes, so I can see usecases for it.

I explicitly pointed out in that section that one of the short-comings of this approach is that it doesn't update and to look at the property-section for an approach where it does (which I still have to write).